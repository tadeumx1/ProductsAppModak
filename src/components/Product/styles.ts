import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
  card: {
    marginBottom: 16,
    borderRadius: 8,
    elevation: 3,
  },
  coverImage: {
    width: 64,
    height: 64,
    marginRight: 10,
    borderRadius: 25,
  },
  content: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  thumbnail: {
    marginRight: 16,
  },
  infoContainer: {
    flex: 1,
    justifyContent: 'center',
  },
  title: {
    fontSize: 16,
    fontWeight: 'bold',
    marginBottom: 4,
  },
  price: {
    fontSize: 14,
    color: 'gray',
  },
});
