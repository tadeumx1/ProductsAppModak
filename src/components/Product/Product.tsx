import React from 'react';
import { View } from 'react-native';
import { Card, Text } from 'react-native-paper';
import FastImage from 'react-native-fast-image';
import { styles } from './styles';

interface ProductProps {
  title: string;
  price: number;
  thumbnail: string;
}

const Product: React.FC<ProductProps> = ({
  title,
  thumbnail,
  price,
}: ProductProps) => {
  return (
    <Card style={styles.card}>
      <Card.Content style={styles.content}>
        <FastImage
          style={styles.coverImage}
          source={{
            uri: thumbnail,
            priority: FastImage.priority.normal,
          }}
          resizeMode={FastImage.resizeMode.contain}
          testID="image-thumbnail-container"
        />

        <View style={styles.infoContainer}>
          <Text style={styles.title}>{title}</Text>
          <Text style={styles.price}>${price.toFixed(2)}</Text>
        </View>
      </Card.Content>
    </Card>
  );
};

export default Product;
