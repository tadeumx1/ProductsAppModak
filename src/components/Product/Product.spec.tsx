import React from 'react';
import { render } from '@testing-library/react-native';
import Product from './Product';

jest.mock('react-native-fast-image', () => {
  const React = require('react');
  const { Image } = require('react-native');

  const FastImage = ({ source, style }: any) => (
    <Image
      source={{ uri: source.uri }}
      style={style}
      testID="image-thumbnail-container"
    />
  );

  FastImage.priority = {
    normal: 'normal',
  };

  FastImage.resizeMode = {
    contain: 'contain',
  };

  return FastImage;
});

describe('Product Component', () => {
  const productProps = {
    title: 'Test Product',
    price: 19.99,
    thumbnail: 'https://example.com/image.jpg',
  };

  it('render the product title', () => {
    const { getByText, toJSON } = render(<Product {...productProps} />);

    expect(getByText('Test Product')).toBeDefined();

    expect(toJSON()).toMatchSnapshot();
  });

  it('render the product price in correct format', () => {
    const { getByText, toJSON } = render(<Product {...productProps} />);

    expect(getByText('$19.99')).toBeDefined();

    expect(toJSON()).toMatchSnapshot();
  });

  it('render the product thumbnail image', () => {
    const { getByTestId, toJSON } = render(<Product {...productProps} />);

    const image = getByTestId('image-thumbnail-container');

    expect(image.props.source.uri).toBe(productProps.thumbnail);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should apply correct styles to the components', () => {
    const { getByText, toJSON } = render(<Product {...productProps} />);
    const title = getByText('Test Product');
    const price = getByText('$19.99');

    expect(title.props.style).toEqual(expect.any(Object));
    expect(price.props.style).toEqual(expect.any(Object));

    expect(toJSON()).toMatchSnapshot();
  });
});
