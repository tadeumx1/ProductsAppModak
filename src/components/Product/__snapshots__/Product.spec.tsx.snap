// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Component render the product price in correct format 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "rgba(247, 243, 249, 1)",
      "borderRadius": 8,
      "marginBottom": 16,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="card-container-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(247, 243, 249, 1)",
        "borderRadius": 8,
        "elevation": 3,
        "flex": undefined,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="card-container"
  >
    <View
      style={
        [
          {
            "flexShrink": 1,
          },
          undefined,
        ]
      }
      testID="card"
    >
      <View
        borderRadiusStyles={
          {
            "borderRadius": 8,
          }
        }
        style={
          [
            {
              "paddingHorizontal": 16,
            },
            {
              "paddingVertical": 16,
            },
            {
              "alignItems": "center",
              "flexDirection": "row",
            },
          ]
        }
      >
        <Image
          source={
            {
              "uri": "https://example.com/image.jpg",
            }
          }
          style={
            {
              "borderRadius": 25,
              "height": 64,
              "marginRight": 10,
              "width": 64,
            }
          }
          testID="image-thumbnail-container"
        />
        <View
          style={
            {
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginBottom": 4,
                },
              ]
            }
          >
            Test Product
          </Text>
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": "gray",
                  "fontSize": 14,
                },
              ]
            }
          >
            $
            19.99
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Product Component render the product thumbnail image 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "rgba(247, 243, 249, 1)",
      "borderRadius": 8,
      "marginBottom": 16,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="card-container-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(247, 243, 249, 1)",
        "borderRadius": 8,
        "elevation": 3,
        "flex": undefined,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="card-container"
  >
    <View
      style={
        [
          {
            "flexShrink": 1,
          },
          undefined,
        ]
      }
      testID="card"
    >
      <View
        borderRadiusStyles={
          {
            "borderRadius": 8,
          }
        }
        style={
          [
            {
              "paddingHorizontal": 16,
            },
            {
              "paddingVertical": 16,
            },
            {
              "alignItems": "center",
              "flexDirection": "row",
            },
          ]
        }
      >
        <Image
          source={
            {
              "uri": "https://example.com/image.jpg",
            }
          }
          style={
            {
              "borderRadius": 25,
              "height": 64,
              "marginRight": 10,
              "width": 64,
            }
          }
          testID="image-thumbnail-container"
        />
        <View
          style={
            {
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginBottom": 4,
                },
              ]
            }
          >
            Test Product
          </Text>
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": "gray",
                  "fontSize": 14,
                },
              ]
            }
          >
            $
            19.99
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Product Component render the product title 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "rgba(247, 243, 249, 1)",
      "borderRadius": 8,
      "marginBottom": 16,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="card-container-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(247, 243, 249, 1)",
        "borderRadius": 8,
        "elevation": 3,
        "flex": undefined,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="card-container"
  >
    <View
      style={
        [
          {
            "flexShrink": 1,
          },
          undefined,
        ]
      }
      testID="card"
    >
      <View
        borderRadiusStyles={
          {
            "borderRadius": 8,
          }
        }
        style={
          [
            {
              "paddingHorizontal": 16,
            },
            {
              "paddingVertical": 16,
            },
            {
              "alignItems": "center",
              "flexDirection": "row",
            },
          ]
        }
      >
        <Image
          source={
            {
              "uri": "https://example.com/image.jpg",
            }
          }
          style={
            {
              "borderRadius": 25,
              "height": 64,
              "marginRight": 10,
              "width": 64,
            }
          }
          testID="image-thumbnail-container"
        />
        <View
          style={
            {
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginBottom": 4,
                },
              ]
            }
          >
            Test Product
          </Text>
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": "gray",
                  "fontSize": 14,
                },
              ]
            }
          >
            $
            19.99
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Product Component should apply correct styles to the components 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "rgba(247, 243, 249, 1)",
      "borderRadius": 8,
      "marginBottom": 16,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="card-container-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(247, 243, 249, 1)",
        "borderRadius": 8,
        "elevation": 3,
        "flex": undefined,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="card-container"
  >
    <View
      style={
        [
          {
            "flexShrink": 1,
          },
          undefined,
        ]
      }
      testID="card"
    >
      <View
        borderRadiusStyles={
          {
            "borderRadius": 8,
          }
        }
        style={
          [
            {
              "paddingHorizontal": 16,
            },
            {
              "paddingVertical": 16,
            },
            {
              "alignItems": "center",
              "flexDirection": "row",
            },
          ]
        }
      >
        <Image
          source={
            {
              "uri": "https://example.com/image.jpg",
            }
          }
          style={
            {
              "borderRadius": 25,
              "height": 64,
              "marginRight": 10,
              "width": 64,
            }
          }
          testID="image-thumbnail-container"
        />
        <View
          style={
            {
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginBottom": 4,
                },
              ]
            }
          >
            Test Product
          </Text>
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": "gray",
                  "fontSize": 14,
                },
              ]
            }
          >
            $
            19.99
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
